packages/arb-checkpointer/checkpointing/checkpointedmachine.go

```diff
@@ -0,0 +1,255 @@
+/*
+* Copyright 2020, Offchain Labs, Inc.
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*
+*    http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+ */
+
+package checkpointing
+
+import (
+	"context"
+	"github.com/pkg/errors"
+	"math/big"
+	"os"
+	"sync"
+	"time"
+
+	"google.golang.org/protobuf/proto"
+
+	"github.com/offchainlabs/arbitrum/packages/arb-avm-cpp/cmachine"
+	"github.com/offchainlabs/arbitrum/packages/arb-checkpointer/ckptcontext"
+	"github.com/offchainlabs/arbitrum/packages/arb-util/common"
+	"github.com/offchainlabs/arbitrum/packages/arb-util/machine"
+	"github.com/offchainlabs/arbitrum/packages/arb-util/value"
+	"github.com/offchainlabs/arbitrum/packages/arb-validator-core/arbbridge"
+)
+
+var errNoCheckpoint = errors.New("cannot restore because no checkpoint exists")
+var errNoMatchingCheckpoint = errors.New("cannot restore because no matching checkpoint exists")
+
+type CheckpointedMachine struct {
+	*sync.Mutex
```

Comment:
I prefer not using the pointer for `sync.Mutex` since it creates unnecessary indirection

Reply:
that was copied from indexedCheckpointer.go